sql/tests: TestRandomSyntaxGeneration failed [parse/format/parse of CREATE LOGICAL REPLICATION STREAM fails]

sql/tests.TestRandomSyntaxGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/16236874?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/16236874?buildTab=artifacts#/) on release-24.2 @ [742a2251a05088a238fc6eb4c2466dace3cb1bb2](https://github.com/cockroachdb/cockroach/commits/742a2251a05088a238fc6eb4c2466dace3cb1bb2):


```
    rsg_test.go:116: Last executed (4): FETCH IN REFRESH
    rsg_test.go:116: Last executed (5): COPY GROUPING TO STDOUT HEADER QUOTE 'string'
    rsg_test.go:116: Last executed (6): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (7): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (8): UNLISTEN FAMILY . SYNTAX
    rsg_test.go:116: Last executed (9): PREPARE LANGUAGE ( JOIN ARRAY [ 484 ] [ error ) AS OPT PLAN 'string'
    rsg_test.go:116: Last executed (10): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (11): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (12): IMPORT COALESCE ( 'string' )
    rsg_test.go:116: Last executed (13): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (14): COMMENT ON SCHEMA USERS . FOLLOWING IS NULL
    rsg_test.go:116: Last executed (15): UNLISTEN INDEX_AFTER_ORDER_BY_BEFORE_AT . FORCE_INDEX
    rsg_test.go:116: Last executed (16): COPY ident . FALSE . NONE TO STDOUT
    rsg_test.go:116: Last executed (17): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (18): COPY FAMILY . ident ( QUERIES ) TO error
    rsg_test.go:116: Last executed (19): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (20): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (21): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (22): COMMENT ON CONSTRAINT BIT ON NEW_DB_NAME IS NULL
    rsg_test.go:116: Last executed (23): FETCH -228 ident
    rsg_test.go:116: Last executed (24): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (25): PREPARE GREATEST ( GEOMETRYCOLLECTIONZ . FAMILY . ident [ -123 ] [ error ) AS OPT PLAN 'string'
    rsg_test.go:116: Last executed (26): COPY INDEX_BEFORE_NAME_THEN_PAREN . HEADER FROM error
    rsg_test.go:116: Last executed (27): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (28): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (29): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (30): COMMENT ON DATABASE SUBSCRIPTION IS 'string'
    rsg_test.go:116: Last executed (31): COPY FAMILY . NONE . ILIKE TO error
    rsg_test.go:116: Last executed (32): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (33): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (34): COPY ident . COLLATE ( ident ) TO STDOUT WITH DELIMITER PLACEHOLDER
    rsg_test.go:116: Last executed (35): REFRESH MATERIALIZED VIEW CONCURRENTLY INDEX_BEFORE_NAME_THEN_PAREN . DESC WITH DATA
    rsg_test.go:116: Last executed (36): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (37): MOVE RELATIVE -212 ident
    rsg_test.go:116: Last executed (38): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (39): PREPARE PUBLICATION ( FORMAT . RETURNS [ ] [ error ) AS OPT PLAN 'string'
    rsg_test.go:116: Last executed (40): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (41): ROLLBACK TO SAVEPOINT EXPERIMENTAL_RELOCATE
    rsg_test.go:116: Last executed (42): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (43): COMMENT ON TYPE FAMILY . SESSION_USER . FAMILY IS NULL
    rsg_test.go:116: Last executed (44): COPY NOTHING_AFTER_RETURNING . MAXVALUE . COLLATION ( WORK , LOCALITY ) TO error
    rsg_test.go:116: Last executed (45): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (46): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (47): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:116: Last executed (48): REFRESH MATERIALIZED VIEW INDEX_BEFORE_NAME_THEN_PAREN . INTO WITH DATA
    rsg_test.go:116: Last executed (49): FETCH ABSOLUTE -122 MATERIALIZED
    rsg_test.go:119: Currently active statements: map[]
    rsg_test.go:355: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d437d2c847dfedbc4972f231c3331c8e/logTestRandomSyntaxGeneration2762069134
--- FAIL: TestRandomSyntaxGeneration (301.87s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-40612

--- a/pkg/sql/parser/sql.y
+++ b/pkg/sql/parser/sql.y
@@ -4729,7 +4729,7 @@ logical_replication_options:
   } 
 | FUNCTION db_object_name FOR TABLE db_object_name
   {
-     $$.val = &tree.LogicalReplicationOptions{UserFunctions: map[*tree.UnresolvedName]tree.RoutineName{$5.unresolvedObjectName().ToUnresolvedName():$2.unresolvedObjectName().ToRoutineName()}}
+     $$.val = &tree.LogicalReplicationOptions{UserFunctions: map[tree.UnresolvedName]tree.RoutineName{*$5.unresolvedObjectName().ToUnresolvedName():$2.unresolvedObjectName().ToRoutineName()}}
   }
 
 // %Help: CREATE VIRTUAL CLUSTER - create a new virtual cluster

--- a/pkg/sql/parser/testdata/create_logical_replication
+++ b/pkg/sql/parser/testdata/create_logical_replication
@@ -62,3 +62,11 @@ DETAIL: source SQL:
 CREATE LOGICAL REPLICATION STREAM FROM TABLE foo, bar ON 'uri' INTO TABLE foo, bar
                                                 ^
 HINT: try \h CREATE LOGICAL REPLICATION STREAM
+
+error
+CREATE LOGICAL REPLICATION STREAM FROM TABLES (t1, t2, t3) ON 'uri' INTO TABLES (s.t4, t5) WITH OPTIONS (FUNCTION f1 FOR TABLE d.s.t5 , FUNCTION f2 FOR TABLE s.t4, FUNCTION f3 FOR TABLE s.t4, MODE = 'immediate')
+----
+at or near ",": syntax error: multiple user functions specified for table s.t4
+DETAIL: source SQL:
+CREATE LOGICAL REPLICATION STREAM FROM TABLES (t1, t2, t3) ON 'uri' INTO TABLES (s.t4, t5) WITH OPTIONS (FUNCTION f1 FOR TABLE d.s.t5 , FUNCTION f2 FOR TABLE s.t4, FUNCTION f3 FOR TABLE s.t4, MODE = 'immediate')
+                                                                                                                                                                                              ^

--- a/pkg/sql/sem/tree/create_logical_replication.go
+++ b/pkg/sql/sem/tree/create_logical_replication.go
@@ -30,7 +30,7 @@ type LogicalReplicationResources struct {
 
 type LogicalReplicationOptions struct {
 	// Mapping of table name to UDF name
-	UserFunctions   map[*UnresolvedName]RoutineName
+	UserFunctions   map[UnresolvedName]RoutineName
 	Cursor          Expr
 	Mode            Expr
 	DefaultFunction Expr
@@ -107,7 +107,7 @@ func (lro *LogicalReplicationOptions) Format(ctx *FmtCtx) {
 
 		// In order to make tests deterministic, the ordering of map keys
 		// needs to be the same each time.
-		keys := make([]*UnresolvedName, 0, len(lro.UserFunctions))
+		keys := make([]UnresolvedName, 0, len(lro.UserFunctions))
 		for k := range lro.UserFunctions {
 			keys = append(keys, k)
 		}
@@ -121,7 +121,7 @@ func (lro *LogicalReplicationOptions) Format(ctx *FmtCtx) {
 			r := lro.UserFunctions[k]
 			ctx.FormatNode(&r)
 			ctx.WriteString(" FOR TABLE ")
-			ctx.FormatNode(k)
+			ctx.FormatNode(&k)
 		}
 	}
 }
@@ -154,10 +154,10 @@ func (o *LogicalReplicationOptions) CombineWith(other *LogicalReplicationOptions
 	if other.UserFunctions != nil {
 		for tbl := range other.UserFunctions {
 			if _, ok := o.UserFunctions[tbl]; ok {
-				return errors.Newf("multiple user functions specified for table %q", tbl)
+				return errors.Newf("multiple user functions specified for table %s", tbl.String())
 			}
 			if o.UserFunctions == nil {
-				o.UserFunctions = make(map[*UnresolvedName]RoutineName)
+				o.UserFunctions = make(map[UnresolvedName]RoutineName)
 			}
 			o.UserFunctions[tbl] = other.UserFunctions[tbl]
 		}

